rust/lance/src/index/vector/hnsw.rs

```diff
@@ -0,0 +1,267 @@
+        ivf::storage::IVF_PARTITION_KEY,
+        Query, DIST_COL,
+    },
+    Index, IndexType,
+};
+use lance_io::traits::Reader;
+use lance_linalg::distance::MetricType;
+use roaring::RoaringBitmap;
+use serde_json::json;
+use snafu::{location, Location};
+use tracing::instrument;
+
+#[cfg(feature = "opq")]
+use super::opq::train_opq;
+use super::VectorIndex;
+use crate::index::prefilter::PreFilter;
+
+#[derive(Clone)]
+pub struct HNSWIndex {
+    hnsw: HNSW,
+    storage: Option<Arc<dyn VectorStorage>>,
+    partition_metadata: Option<Vec<HnswMetadata>>,
+}
+
+impl Debug for HNSWIndex {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.hnsw.fmt(f)
+    }
+}
+
+impl HNSWIndex {
+    pub async fn try_new(hnsw: HNSW, reader: Arc<dyn Reader>) -> Result<Self> {
+        let schema = Schema::try_from(&arrow_schema::Schema::new(vec![
+            NEIGHBORS_FIELD.clone(),
+            VECTOR_ID_FIELD.clone(),
+        ]))?;
+
+        let reader = FileReader::try_new_from_reader(
+            reader.path(),
+            reader.clone(),
+            None,
+            schema,
+            0,
+            0,
+            None,
+        )
+        .await?;
+
+        let partition_metadata = match reader.schema().metadata.get(IVF_PARTITION_KEY) {
+            Some(json) => {
+                let metadata: Vec<HnswMetadata> = serde_json::from_str(json)?;
+                Some(metadata)
+            }
+            None => None,
+        };
+
+        Ok(Self {
+            hnsw,
+            storage: None,
+            partition_metadata,
+        })
+    }
+
+    fn get_partition_metadata(&self, partition_id: usize) -> Result<HnswMetadata> {
+        match self.partition_metadata {
+            Some(ref metadata) => Ok(metadata[partition_id].clone()),
+            None => Err(Error::Index {
+                message: "No partition metadata found".to_string(),
+                location: location!(),
+            }),
+        }
+    }
+}
+
+#[async_trait]
+impl Index for HNSWIndex {
+    /// Cast to [Any].
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    /// Cast to [Index]
+    fn as_index(self: Arc<Self>) -> Arc<dyn Index> {
+        self
+    }
+
+    /// Retrieve index statistics as a JSON Value
+    fn statistics(&self) -> Result<serde_json::Value> {
+        Ok(json!({
+            "index_type": "HNSW",
+            "metric_type": self.metric_type().to_string(),
+        }))
+    }
+
+    /// Get the type of the index
+    fn index_type(&self) -> IndexType {
+        IndexType::Vector
+    }
+
+    /// Read through the index and determine which fragment ids are covered by the index
+    ///
+    /// This is a kind of slow operation.  It's better to use the fragment_bitmap.  This
+    /// only exists for cases where the fragment_bitmap has become corrupted or missing.
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        todo!()
+    }
```

Comment:
We might not ever need this.  It was only needed for a migration path because some old versions of the dataset didn't store the fragment bitmap properly.  Since this index is new it's probably ok there is no implementation.  Maybe change to `unimplemented!()` instead of `todo!()`.